### ISSUE TITLE
Public ParserPNG class for more fine grained PNG decoding?

### ISSUE THREAD

**Issue by derselbst** [t0rakka/mango#16]
Hello,

I would like to use mangos PNG decoder in my image viewer. For that I am looking for a more fine grained control of the PNG decoding. In particular I would like to

* determine the rough decoding progress (i.e. get a callback whenever a scanline of the image has been decoded, allowing me to invalidate that part of the view and redraw it), 
* cancel decoding at pretty much any time,
* nice to have: decode the PNG file only partially (e.g. skip a scanlines from top and bottom).

To achieve this, the ParserPNG class would probably need to be made public and potentially making member functions like `process()` and `process_range()` virtual, allowing them to be overridden and customized by a downstream application.

Also I'm uncertain if inheritance should be used to address this, or whether my app should implement a callback struct which then would be regularly called by ParserPNG class.

What's your opinion about those points?

**Comment by t0rakka** [comment]
The process_range is called in tasks in the threadpool so the callbacks might even come in the "wrong order".  The good news is that there is cancel() member function in the ConcurrentQueue but the bad news is it won't cancel tasks that have already started that would be something the task itself would have to know how to do; basically a flag that is checked at whatever is convenient granularity. 

I think that I do get what you want.. you want to be able to really smoothly skim through the files w/o getting blocked until the full image is available (= would feel smooth and not jerky no matter how large file is being opened).I have pre-fetcher in my own app for that kind of use.. basically i have state of sequence of images (ie. folder and it's hierarchy linearizes as list of files. It has different states: not present, loading, processing and ready.. while image isn't ready it displays a symbol and UI is still responsive and not blocked so you can go forward and backward in the list at any time.. for large images it just might take a noticeable time for the image to finally appear.. it's not as good as what could be done if we have  real cancellation and "preview" capability, you might get away by generating thumbnails in your own local cache (I don't recommend storing them with the user images ugh..). thumbnails aren't really a thing with png's and with jpg's they are kind of rare as well.. with jpeg as with png you still need to decode the whole bitstream until you get something on screen unless the storage is progressive (jpg) or interlaced (png) but those modes come with performance penalties, for instance, multiple passes over the data is sub-optimal and the streams are larger so more data to process and so on.

Now to actionable stuff. I think the ImageDecodeOptions is the place where user probably should provide callbacks or interface to control decoding externally.. (see include/image/decode.hpp). I can't say what it might look like, but I don't like polling so it might have to involve messages/event in one way or other. For now I will just say  will think about it.

**Comment by t0rakka** [comment]
Also note for image viewing application: it requires sanitation and strictness order of magnitude more than what I am doing right now, the primary use case I had in mind for this was application development, mostly graphics related where the files are generally hand-picked and tested by the application developers/designers. I have fuzzed the code to a degree and it doesn't crash on the first broken png or jpg file but it could be better. Contributions are welcome, this is just a hobby project these days I don't develop this full time, more like hour here and other there that's why I don't go around advertising and often steer people towards more battle-tested stuff like libjpeg-turbo and libpng, they have massive user base and have gone through massive catalogue of image files.. that said I am interested improving the code at any time. ;)

**Comment by derselbst** [comment]
> I think that I do get what you want.. you want to be able to really smoothly skim through the files w/o getting blocked until the full image is available

Exactly. I'm not using prefetching though - it's really just about not getting blocked. You can try it if you're interested: https://github.com/derselbst/ANPV

> I think the ImageDecodeOptions is the place where user probably should provide callbacks

I have considered this as well. The cons are that it's not supported by all interfaces. E.g. openJpeg or JpegXL do not implement that, hence your decoder cannot make use of it in general. That's why I was preferring the ParserPNG, since that's all I need at the moment.

I'm aware, that potential callback might occur concurrently - that's fine.

> Contributions are welcome

Ok, great. I'll look into getting it working with the ParserPNG class, probably gonna take a few weeks as well.

**Comment by t0rakka** [comment]
Yes, okay, after sleeping on it I came to conclusion that is what we want. I think it would be reasonable that JPG and PNG at least have support for it, those are the formats that image viewers will most of the time deal with. I will do some prototyping and testing soon-ish, I am kind of glad that have the decoder object instance that makes the synchronization so much nicer to deal with.. the destructor will guarantee nothing is left hanging.

- reporting progress: yes
- cancellation: yes
- preview/progressive decoding: need to work on it to know how feasible that is

Doing progressive updates always is a definite no since it will slow down decoding substantially, so it's really bad when not wanted or needed. When it's requested that is another story: then we should provide some kind of fast preview that is updated realtime but what it could be depends on the type of image. If PNG is not interlaced what we can do is make the ranges available when we know they are resolved so image can "fill" in gradually. With interlaced PNG we can provide full image that becomes more refined as decoding progresses. What I want is that how the data is sent to user is just one way so that there is incentive to use it, in other words write the code once and it keeps working.

How that could work is that the target surface user gives to the decoder is written by the decoder and blanks are filled so that any time the user gets notification that the destination is refreshed it is possible to use as-is, so it must be cleared initially to black so that there is no garbage in the blank areas at the very minimum. Interlaced decoder could fill the blanks with same color so it just that resolution improves on each update. Progressive JPG can first have MCU with same color at first, and then as it gets more frequencies decoded the resolution improves. The non-progressive images would just fill the black initial void with more pixels as they become available.

Currently the progressive JPG decoder decompresses all passes for MCU and does the color processing for whole image when it has all of the data available; doing it after each pass would be wasteful, but in preview-wanted scenario we would waste a bit of CPU to get the smooth updates rolling. This would be configured in the Options and default would be single-shot decoding as it is now.

We could also make the ranges (for JPG) smaller when in preview-wanted mode so that updates are more frequent, maybe even make it configurable we'll see what works best after coding a little bit. With PNGs the ranges are pre-determined at the time the PNG is encoded, to start early processing of decoded data would need to know where the deflate decoder is at and that's a bit too much as would prefer to not touch code that comes downstream for me.

I might be over-thinking this, let's see how it works out.. maybe getting draft in some experimental branch as quickly as possible would be best so that you can check it out and refine the code until we're happy what's going on. Stay tuned. :)

**Comment by t0rakka** [comment]
`
class ImageDecodeController
{
virtual void update(const State& state) = 0;
void cancel();
};
`

Something like that, user can derive from it, and the decoder calls the update member function when something actionable happens like decode target surface has fresh data. The State could inform what regions are updated if it's just range update or whole surface if it is progressive or interlaced update. Other relevant information could also be communicated to the user through this. The cancel could also be in the controller and that call would be blocking until the decoder is done.

The decode() is currently blocking. The controlled variant needs to be non-blocking, but I kind of like the blocking in decode() it's too convenient. Perhaps decode() that takes ImageDecodeController as argument and returns immediately. It's safe to delete the decoder only once controller has update call where state indicates that the decoding is complete. The decoder destructor still blocks so that's the final barrier where all threads involved must be complete and all queued tasks also must be complete. This way user don't NEED to wait for the update+state.complete but it is still done for the user convenience so that he can do what he must, the thread and memory management is still in the decoder.

Yeah.. that would work.. I have similar arrangement for GPU accelerated JPG decoding where I just leverage the existing parser and pass the data to external decoder. I wrote a prototype but it can only decode certain form of JPEG and is a bit crap, but as a concept it worked.. put it on another branch where it is parked but that is another story. I wouldn't look into it it's not usable as it is but learned from it. ;) This is a bit different since almost everything still happens internally in the decoder it's just a different MODE or decoding, let's call it "asynchronous user-controllable decoding mode"

**Comment by t0rakka** [comment]
One thing that comes to mind is that update() should return a snapshot of the target surface, we don't want to get something that was being written into while user reads from it. This implies that we must make copies of the data so that's extra performance hit but I'll sort out the details. We must make a copy since even if we have internal buffer where we decode and only blit to target before calling update we can't guarantee user has consumed the data fully before we update it again. A copy or snapshot must be made and sent to the user that is isolated from what decoder is doing; it can't write over the data it has passed to user. Yeah, I think we have an idea of design now.

**Comment by derselbst** [comment]
Thanks for your thoughts. Let me just clarify: I don't need progressive decoding update (like the one that can be observed in ANPV when opening a progressive JPEG). I just need a callback that informs my app "this region has finished decoding, you might want to redraw that ROI" - whereas ROI will probably mean entire scanlines in practice.

> so it must be cleared initially to black so that there is no garbage in the blank areas at the very minimum.

Transparent black ideally, yes, but `calloc()` will take care of this.


> class ImageDecodeController { virtual void update(const State& state) = 0; void cancel(); };

That looks good. How about changing the cancel() to a `throwIfCanceled()`, that your decoder just calls in some places where it makes sense (loops, etc) and the app can just throw an exception if a flag from somewhere has been set? The exception will go all the way through mango and will be catched by the app.

> Perhaps decode() that takes ImageDecodeController as argument and returns immediately

I wouldn't change any of the existing calling conventions for decode(), they are fine in my opinion. The ImageDecodeController struct could be passsed via the previously mentioned `ImageDecodeOptions`, I'd guess.

> we don't want to get something that was being written into while user reads from it. This implies that we must make copies of the data so that's extra performance hit

Oh no, no copy pls. I'm dealing with big images. The callback really shall only happen, when the final pixels are written to the decode surface buffer and never looked at it again by the decoder. The decode buffer is allocated and owned by the user's app anyhow.

**Comment by t0rakka** [comment]
There is async_decode branch where I done some clean up to the decode interface, probably will merge to main before continuing further. Added a member function to ConcurrentQueue to ask if it was cancelled; this allows tasks to cancel themselves (the tasks must capture queue they belong and then can ask the queue if it was cancelled). 

The controller idea has another option; AsyncImageDecoder which inherits from the ImageDecoder and has the event notify member function user can override. We'll see where it goes; it must be practical / convenient to set up and use.

**Comment by t0rakka** [comment]
If we just update parts of the image like scanline or range, ROI works. If it is progressive image where resolution improves when more passes of progressiveness are resolved then it will always mean the whole surface is touched. That's where the concurrency can mess up reading since there is no isolation of regions like the ROI/crop rectangle approach does; what is indicated as updated is going to be final. With progressive the whole image is in "fluid" state.

The "percentage completed"  will be N / Nmax, where Nmax is a value that is based on decoding mode.. let's say it's scanlines, if image has 1080 scanlines the progress made is N / 1080, user can calculate percentage or whatever he wants from that.  If image is split into ranges, then Nmax is number of ranges. If it's progressive JPEG, the progressive scans are indicated by AC range which is from 1 to 63, the value 0 is always DC.. so progressive progress is indicated by highest AC (and DC) index the pass has processed. At first DC is resolved so that's 1/64, then let's say 7 DC's are processed we're at 8/64 and when at 64/64 the progressive decoding is complete. This scheme works with all PNG and JPG modes. Tiled image formats won't fit this scheme but they can just use Nmax = nr of tiles.

**Comment by derselbst** [comment]
> There is async_decode branch where I done some clean up to the decode interface, probably will merge to main before continuing further. Added a member function to ConcurrentQueue to ask if it was cancelled;

Ok, sounds good to me, we can try that.

> If we just update parts of the image like scanline or range, ROI works. If it is progressive image where resolution improves when more passes of progressiveness are resolved then it will always mean the whole surface is touched. That's where the concurrency can mess up reading since there is no isolation of regions like the ROI/crop rectangle approach does; what is indicated as updated is going to be final. With progressive the whole image is in "fluid" state.

I'd vote to keep it simple and move progressiveness out of scope. While this works quite ok for JPEG, I tried it for interlaced PNGs in the past and found that those "intermediate-downsampled decoding results" are not usable / not nice to look at, because they have their colors pretty messed up. And as you described, calculating a progress will become more complicated in this case as well.

**Comment by t0rakka** [comment]
Started prototyping; got first update message from the JPEG decoder. I have ImageDecodeListener that has the callback but it's life-time has to be carefully managed; it must be created before the decoder so that decoder destructor is ran first, if the listener disappears while decoding is still in progress that's of course not very good.

Here's the basic arrangement: when the decode is done in the higher-level ImageDecoder, it picks up the async requirement from the options.. and launches a thread and calls the decode as usual. Then the destructor joins the thread which blocks until the thread returned; this manages the ImageDecoderInterface life-time as it's stored in the higher-level ImageDecoder.

It would be more straightforward if the callback object is managed in the decoder as well, this way they are always destructed in the correct order. The simplest solution is that the ImageDecoder is inheritable by the user, so and if the update() is pure virtual it must be implemented so it will be pretty hard to screw up. It's at this point tempting to also create the decoding target surface (Bitmap) in the ImageDecoder, but we don't want this to be default (obviously) so we inherit AsyncImageDecoder from the ImageDecoder, then the user inherits from that.. and gets the update notifications and the decoding target allocated in correct format ; this means direct decoding always happens which means more speed. The downside is that you can't map a GPU texture and decode there directly.. so the auto-allocated target storage is not a must-have feature.

The example code is in async_decode.cpp in under examples/image and it looks pretty crap at the moment but got the first update message from async decoder, that's a step forward. This example is the testing ground to make it nicer to use, plus will add cancellation and some measure of cancellation latency at different times vs. full decoding, etc. WIP.

**Comment by t0rakka** [comment]
Forgot to mention that if we have ImageDecoder -> AsyncImageDecoder -> UserDecoder, this means the AsyncImageDecoder can make all the arrangements needed to make the decoding async, no need to fiddle with options and Listener crap disappears.

**Comment by t0rakka** [comment]
.. and life-time of about everything would be automatically managed so user can't screw up so easily. The decoding target surface is still open issue, having it on the Async decoder would eliminate two problems: life time and direct decoding.

Worth mentioning again: async decode will not block, but the destructor will, but this leaves open interval while decoding is happening where user gains control (gets updates, and option to cancel).

**Comment by t0rakka** [comment]
Also, currently codecs that don't support async still can get launched (with a warning message), and they are still caught at the destructor. The timings are ~same as without async but it's extra thread + the wait at join in dtor. If I post a completion update once the decode() returns from non-async decoder interface it would work just like async-capable decoder just no cancel or updates (except completion).

A small beauty issue with that is that when user creates AsyncImageDecoder of his own, he will only know if it is async capable or not after creating it, since this is asked from the internal interface which is created when the decoder is created.. this is where the before-mentioned detail that non-async decoder can still be async just with less control options. I'd go with that initially and probably in a long term as well.

**Comment by t0rakka** [comment]
It still has some value that the decoding is non-blocking even if not controllable, more options to the user.

**Comment by t0rakka** [comment]
Prototype JPEG async decoder; supports cancel and updates. Not all modes yet update fully and there are some gaps here and there but it proves the idea is sound. The cancellation is also very crude but if I launch decoder, sleep for 3 ms, the completion comes in 3-4 ms, so it's fairly responsive. The test image is 28 ms and can cancel it in 0 ms if cancel immediately.

The updates don't work when there is clipping or color conversion as the decoding goes into the temporary surface. I think this is a good opportunity to make the conversion inline with the decoder so that the target pixels are always what the update indicates. After the JPEG is in somewhat shape will do the PNG.

The API isn't yet what I hoped it would be so expect very rough changes coming in.

**Comment by t0rakka** [comment]
Didn't have to do much trickery with the cancel checking; it's atomic bool but since don't check it for each MCU the cost disappears entirely. Checking queued tasks doesn't do much so keeping that code as-is so that it doesn't become too complicated; 1 ms latency for cancel is alright (if even that). If found out it needs tuning later then so be it.

**Comment by jliimatt** [comment]
JPEG async is now functional. The API could still be nicer. This is obvious but the update() calls synchronization is up to the user as the calls can happen in any thread and it is even possible for multiple calls to come at the same time. The callback is currently a ImageDecodeRect: x, y, width, height. The progress isn't yet implemented, coming soon. There is a separate complete callback when the decoding is finished.

The backend/decoder interface side is fairly nice to write and maintain so far..

**Comment by jliimatt** [comment]
Update: the AsyncImageDecoder::launch() creates a thread, starts decoding there and destructor joins the thread; while the decoding is taking place the cancel() can stop it, and setting listener forwards the update calls to it.

This works but the thread creation at launch is just for convenience; it doesn't really matter if the decode() is blocking; what we want is that we receive updates: check, it happens regardless of which thread we're at. Cancellation is difficult FROM THE SAME THREAD if decode() is blocking, so here's the latest: what if the decode remains blocking and we just provide the way to receive updates the cancel from another thread. We need object for this: the listener and async decoder can be fused going back to the original controller design.

The machinery itself is in the ImageDecoderInterface ; it listens for cancellation and sends updates that part turned out OK so far, the JPEG prototype works nicely. So the API quest still continues on the user-facing side and is just using the "Interface" as the engine. 

The side-effect of this is that the color conversion and clipping (when in non-direct decoding mode) is now in the worker threads and more local to the data so some upgrades occurred.

**Comment by derselbst** [comment]
Thanks for keeping me posted and putting so much effort into this already! I hope to find some time on the weekend to have a closer look into this.

**Comment by t0rakka** [comment]
One issue.. when decoding mango-encoded JPEGs the updates come so fast that you might not be able to do anything with the updates before the decoding is complete. :_D   I have this 2560x1600 image that decodes in 5 ms, you'll be busy doing anything with the slices of decoded image as they keep coming. A positive problem, I guess.

**Comment by t0rakka** [comment]
But, if your use case is really big images then it's a different story.. I need bigger test pics.. :D

**Comment by t0rakka** [comment]
It's starting to get shape. The callback is a lambda so no need to fiddle around with overrides. The async launch returns future<ImageDecodeState> so for synchronization you just do .get() on it and get the decode state as extra bonus. You can also leave it hanging and the destructor will synchronize but it's not as nice since you have to control the scope.

I could detach the thread as well, it captures the ImageDecodeInterface (renamed!) so the AsyncImageDecoder can go out of scope if you don't need it for cancel(). I have to consider the implications but we're getting there.

**Comment by derselbst** [comment]
The new async interface is very handy and it works really well! Thanks a lot!

The roi updates work great for JPEG already. There is a tiny issue: Currently, the roi update may exceed the surface dimensions. I think you need to take `yblock_last` into account here:
https://github.com/t0rakka/mango/blob/56383ebdd164707cc9e40133c4b91baa96b4fa4f/source/mango/jpeg/jpeg_decode.cpp#L2098

Or alternatively you could also document that the roi may exceed image dimensions and that the user should crop it if required.

If you need a test JPEG, you could use the one I've provided 4 years ago in my first bug report. Here it is again:
https://drive.google.com/file/d/1DO3fiQipH13ER4nzDUIsgNjJDRHOf56k/view?usp=sharing

I'm looking forward to PNG async support :) In case you need a big PNG test image you can use the one from #3. Let me know if you need additional files.

> I could detach the thread as well, it captures the ImageDecodeInterface (renamed!) so the AsyncImageDecoder can go out of scope if you don't need it for cancel(). 

I have no preference here, sry. I'd say the current behavior is just fine.

**Comment by t0rakka** [comment]
No, I must crop the update rect so that it's correct. =D  The PNG decoder been refactored heavily to enable this, the parallel segment decoding is fully supported. I record the IDAT and fDAT now in vector<ConstMemory> but still temporary make a concatenated buffer to decode in one-shot, but been slowly making it possible to decode IDAT-at-a-time so that can have that sweet progressiveness with default-encoded files as well, it's not going to be as parallel since the "filter bytes" for each scanline still block that.. things must be done in specific order which is a bummer but at least cancellation should be better when we can cancel in middle of the decompression. Still on-going..

**Comment by t0rakka** [comment]
`                rect.height = std::min(y1 * yblock, height) - y0 * yblock;
`
That should do the trick, it's in 4 places..

**Comment by t0rakka** [comment]
- The non-pLLD decoding still needs optimization and progressiveness.
- Test file suite passes (all types of decoding and corner case tests)
- Oops: didn't test all combinations when target format doesn't match, but should be covered
- temporary decode surface management needs cleaning up

**Comment by t0rakka** [comment]
PNG decompression is now done IDAT a a time so cancellation has fine granularity. The filtering and color conversion is still after decompression in it's own pass. I'll fuse it next with the decompression so that updates can be done frequently.

I detect the progressive mode from the callback pointer; if it isn't set we still use the faster old-fashioned one-shot decoding, the progressiveness costs a little so turn it on only when needed.

**Comment by derselbst** [comment]
Pretty amazing what you've achieved over the past few days!

PNG ROI updates work really well! So does the cancellation for both, JPEG and PNG.

I've found a minor issue with the ROI updates: For a progressive JPEG image and when multihreaded decoding is disabled, there will be only a single ROI update for the entire image at the very end, which occurs just before `process_range()` completes called from here:
https://github.com/t0rakka/mango/blob/e721e07a082e04e40dee26d0c504460f31541c57/source/mango/jpeg/jpeg_decode.cpp#L2598

Would it be possible to improve this a bit?

**Comment by t0rakka** [comment]
Unfortunately not; the progressive jpeg's deal with whole image; at first the MCU's get 1 DC, it can be used to get low-res version of the color image fast; it's like every sample in 8x8 block is same color. There are 63 AC's that are transmitted progressively and add detail so the 8x8 blocks become more refined. The progressive update would always indicate whole image. For the update to make sense we would have to resolve the colors as many times there are updates; very wasteful.

The first DC update only contains less than 2% of the payload as it's only 1/64th of the pixels so one color resolve after that would make sense when in "async mode" (ie. the callback pointer is set), it would indicate full image and another one once the decoding is complete. If that works out maybe we could add more updates at increased decoding time cost. Maybe something for the ImageDecodeOptions later. I also delay the updates for interlaced PNGs the same way; they are only resolved once all of the data is decoded.

Before those I will validate the latest code before merging and then we'll see where it goes.

**Comment by t0rakka** [comment]
btw. I didn't add progress meter as it can be calculated by the user if he keeps track of the nr. of scanlines updated, say you get rect.height, just sum these up and divide by image height to get the fraction. This wouldn't work with multiple full-image updates I guess.

**Comment by t0rakka** [comment]
Why I didn't add it was that the updates can arrive "out of order" because of the parallel nature of things happening, so it would be a bit lame to get 40% done BEFORE 10% done (just an example). One thing that could work is the update contains fraction of work it does and you sum them up in the receiving end, this way the full image updates wouldn't matter.

**Comment by t0rakka** [comment]
float progress; // how much "progress" this update contributes :D

**Comment by derselbst** [comment]
Ok, no problem.

> btw. I didn't add progress meter as it can be calculated by the user if he keeps track of the nr. of scanlines updated

That's fine, it's exactly how  I do it.

**Comment by t0rakka** [comment]
Don't be surprised if the "progress" appears in near future, it will make the update future-proof once interlaced/progressive-jpeg get upgraded ROI support. If it's user setting in options that would make it fast by default and you could opt-in for more updates. The biggest worry for me is that with large images writing or reading the whole image is going to be expensive so not sure if it is worth the hassle.

The async decoding has some hidden cost that I don't understand where it comes from; could it just be that expensive to communicate between threads, or, it could be the future-get, I'll drill into it to find if there is something to fix or improve. I'll need the traces to look at to find out if there is anything obvious. It could just be a thousand cuts scenario.

**Comment by t0rakka** [comment]
Fixed one annoying thing; the non-async decoders don't update at all, so made it so that they now callback full image once decoding is complete. This simplifies code elsewhere as it has to do less "what if, then if.." 

In a long run more decoders will get cancellation/progressiveness support.. also looked at the decoding speeds, there is still work left in that area.

**Comment by t0rakka** [comment]
I will merge this soon as it passes all tests..

**Comment by derselbst** [comment]
> the non-async decoders don't update at all, so made it so that they now callback full image once decoding is complete

That's great! For my use case it's working perfectly fine now, thanks a bunch!

**Comment by t0rakka** [comment]
MANGOPOCALYPSE II: The Merging Day is coming soon..

**Comment by t0rakka** [comment]
Done! I close this now, scrolling down here is becoming real work.
